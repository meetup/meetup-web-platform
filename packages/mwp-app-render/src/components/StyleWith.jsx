import React from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';

/**
 * @module StyleWith
 *
 * Adds style module CSS to `head` for both
 * server and client render via `Helmet`.
 *
 *
 * Usage:
 * ```
 * import StyleWith from 'mwp-app-render/lib/components/StyleWith';
 * import styles from './myComponent.module.scss';
 *
 * const MyComponent = (props) => {
 *		return (
 *			<StyleWith styles={[styles]}>
	 *			<p className={styles.someClass}>Hello world!</p>
 *			</StyleWith>
 *		);
 * };
 * ```
 */
const StyleWith = props => {
	const { styles, children } = props;

	const moduleCSS = new Set();
	styles.forEach(style => {
		moduleCSS.add(style._getCss()); // eslint-disable-line no-underscore-dangle
	});

	return (
		<div>
			<Helmet defer={false}>
				<style type="text/css">
					{[...moduleCSS].join('')}
				</style>
			</Helmet>
			{children}
		</div>
	);
};
StyleWith.propTypes = {
	styles: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default StyleWith;
