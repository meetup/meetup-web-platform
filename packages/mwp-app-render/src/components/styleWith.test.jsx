import React from 'react';
import { shallow } from 'enzyme';
import StyleWith from './StyleWith';

const MOCK_PROPS = {
	styles: [
		{ _getCss: () => '.testClass1 {color: inherit;}' },
		{ _getCss: () => '.testClass2 {page-break-after: avoid;}' },
	],
};

describe('StyleWith', function() {
	it('renders with mock props', function() {
		expect(shallow(<StyleWith {...MOCK_PROPS} />)).toMatchSnapshot();
	});
});
