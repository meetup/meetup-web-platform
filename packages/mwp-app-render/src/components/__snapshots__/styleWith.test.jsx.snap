// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleWith renders with mock props 1`] = `
<div>
  <HelmetWrapper
    defer={false}
    encodeSpecialCharacters={true}
  >
    <style
      type="text/css"
    >
      .testClass1 {color: inherit;}.testClass2 {page-break-after: avoid;}
    </style>
  </HelmetWrapper>
</div>
`;
